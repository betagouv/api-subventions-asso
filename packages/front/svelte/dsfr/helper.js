export const getIconClass = icon => {
    if (!icon) return "";
    else return icon.startsWith("fr-icon-") ? icon : `fr-icon-${icon}`;
};
