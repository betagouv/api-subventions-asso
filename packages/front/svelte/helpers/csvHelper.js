export const linesToCsv = array => array.join("\n");

export const arrayToLine = array => array.join(";");

export const downloadCsv = (content, filename) => {
    const encodedUri = encodeURI("data:text/csv;charset=utf-8," + content);
    const link = document.createElement("a");
    link.setAttribute("href", encodedUri);
    link.setAttribute("download", `${filename}.csv`);
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
};

export const buildCsv = (headers, rows) => {
    if (!Array.isArray(headers) || !Array.isArray(rows)) return undefined;
    return linesToCsv([arrayToLine(headers), ...rows.map(arrayToLine)]);
};
