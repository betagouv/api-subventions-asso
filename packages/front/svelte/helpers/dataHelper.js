export const valueOrHyphen = value => value || "-";
