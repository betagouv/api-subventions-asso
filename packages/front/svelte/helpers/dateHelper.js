import { capitalizeFirstLetter } from "./textHelper";

export const sortByDateAsc = (a, b) => {
    if (!a.date) return 1;
    if (!b.date) return -1;
    return a.date - b.date;
};

export const withTwoDigitYear = date => {
    if (date instanceof Date) {
        const localeDate = date.toLocaleDateString();
        const last2Digit = localeDate.slice(-2);
        const dateWithoutYear = localeDate.slice(0, 6);
        return dateWithoutYear.concat(last2Digit);
    } else return date;
};

export const dateToDDMMYYYY = _date => {
    const date = new Date(_date);
    return `${getDayWithZero(date.getDate())}/${getMonthWithZero(date.getMonth() + 1)}/${date.getFullYear()}`;
};

export const getDayWithZero = day => ("0" + day).slice(-2);

export const getMonthWithZero = day => ("0" + day).slice(-2);

export const isValidDate = date => date instanceof Date && !isNaN(date);

export const getOneYearBeforeDate = date => {
    const prevDate = new Date(date);
    prevDate.setFullYear(date.getFullYear() - 1);
    prevDate.setDate(date.getDate() + 1);
    return prevDate;
};

export const YEAR_START_LOGS = 2021;

export const YEAR_CHOICES = [];
for (let year = YEAR_START_LOGS; year <= new Date().getFullYear(); year++) YEAR_CHOICES.push(year);

export function monthCapitalizedFromId(monthId) {
    const monthStr = new Date(Date.parse(monthId + 1 + " 1, 2022")).toLocaleDateString(`en`, { month: "long" });
    return capitalizeFirstLetter(monthStr);
}
