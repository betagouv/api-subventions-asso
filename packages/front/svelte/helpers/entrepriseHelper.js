export function isAssociation(code_juridique) {
    const LEGAL_CATEGORIES_ACCEPTED = [
        // TODO: créer un package shared pour partager les cat juridique entre le front et le back
        "9210", // Association non déclarée
        "9220", // Association déclarée
        "9221", // Association déclarée d'insertion par l'économique
        "9222", // Association intermédiaire
        "9223", // Groupement d'employeurs
        "9230", // Association déclarée, reconnue d'utilité publique
        "9240", // Congrégation
        "9260", // Association de droit local (Bas-Rhin, Haut-Rhin et Moselle)
        "9300", // Fondation
        "92" // Association loi 1901 ou assimilé
    ];

    return LEGAL_CATEGORIES_ACCEPTED.includes(code_juridique);
}
