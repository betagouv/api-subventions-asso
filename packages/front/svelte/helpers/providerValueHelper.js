// TODO: make this file available to API and FRONT ? => Create a core/shared package ? Put it in DTO package because PV is related to DTO ?

export const flatenProviderValue = providerValueObject => {
    const reduceProviderValues = (acc, prop) => {
        return {
            ...acc,
            [prop]:
                isProviderValues(providerValueObject[prop]) || isProviderValue(providerValueObject[prop])
                    ? getValue(providerValueObject[prop])
                    : flatenProviderValue(providerValueObject[prop])
        };
    };

    if (["number", "string"].includes(typeof providerValueObject) | !providerValueObject) return providerValueObject;

    if (Array.isArray(providerValueObject)) {
        return providerValueObject.map(ob => flatenProviderValue(ob));
    }

    const providerValueObjectProps = Object.keys(providerValueObject);
    return providerValueObjectProps.reduce(reduceProviderValues, {});
};

export const getValue = pv => {
    return getPropFromProviderValue("value")(pv);
};

export const getProvider = pv => {
    return getPropFromProviderValue("provider")(pv);
};

export const getDate = pv => {
    return getPropFromProviderValue("last_update")(pv)?.toString();
};

export const getPropFromProviderValue = value => data => {
    if (!data) return undefined;
    if (Array.isArray(data)) {
        if (data.length === 0) return undefined;
        return data[0][value];
    }
    return data[value];
};

export const isProviderValue = value => {
    return !!value && typeof value == "object" && "value" in value && "provider" in value;
};

export const isProviderValues = values => {
    return Array.isArray(values) && !!values.filter(v => v).length && values.every(v => isProviderValue(v));
};
