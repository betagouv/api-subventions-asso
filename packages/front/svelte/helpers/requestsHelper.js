export function toQueryString(queryObj) {
    return Object.entries(queryObj)
        .reduce((query, [key, value]) => {
            if (value === undefined) return query;
            return query.concat(`${key}=${value}`);
        }, [])
        .join("&");
}
