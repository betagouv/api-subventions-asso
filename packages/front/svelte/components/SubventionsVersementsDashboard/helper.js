import lodash from "lodash";
import { sortByDateAsc, isValidDate } from "../../helpers/dateHelper";

const isVersementValid = versement => {
    return isValidDate(new Date(versement.dateOperation)) && typeof versement.amount == "number";
};

const linkVersementsToSubvention = elements =>
    elements.reduce((acc, group) => {
        const subventions = group.filter(element => element.isSub);
        const versements = group.filter(element => element.isVersement && isVersementValid(element));

        const lastSub = subventions.reduce((lastSub, curr) => {
            if (!lastSub) return curr;

            if (lastSub.date_fin < curr.date_fin) return curr;
            return lastSub;
        }, null);

        const siret = lastSub?.siret || versements.find(v => v.siret)?.siret;
        acc.push({
            subvention: lastSub,
            versements: versements,
            siret,
            date: lastSub?.annee_demande ? new Date(lastSub.annee_demande) : getLastVersementsDate(versements),
            year: lastSub?.annee_demande ? lastSub.annee_demande : getLastVersementsDate(versements).getFullYear()
        });

        return acc;
    }, []);

export const mapSubventionsAndVersements = ({ subventions, versements }) => {
    const taggedSubventions = subventions.map(s => ({ ...s, isSub: true }));
    const taggedVersements = versements.map(s => ({ ...s, isVersement: true }));

    const elementsGroupedByVersementKey = [...taggedSubventions, ...taggedVersements].reduce(groupByVersementKey, {
        none: [],
        withKey: {}
    });

    const flatenElements = [
        ...Object.values(elementsGroupedByVersementKey.withKey),
        ...elementsGroupedByVersementKey.none
    ];

    const uniformizedElements = linkVersementsToSubvention(flatenElements);
    return uniformizedElements.sort(sortByDateAsc);
};

// Return year of subvention or versement as a string
export const getYearOfElement = element => {
    if (element.isSub) return getSubventionYear(element);
    if (element.isVersement) return getVersementYear(element);
};

export const getSubventionYear = subvention => subvention.annee_demande;

export const getVersementYear = versement => {
    if (versement.periodeDebut) return new Date(versement.periodeDebut).getFullYear();
    if (versement.dateOperation) return new Date(versement.dateOperation).getFullYear();
};

const groupByVersementKey = (acc, curr) => {
    if (!curr.versementKey) {
        // impossible de lier car pas de clef de liaison
        acc.none.push([curr]);
        return acc;
    }

    const year = getYearOfElement(curr);
    const yearStr = year ? String(year) : "";
    const key = `${curr.versementKey}-${yearStr}`; // Discuter avec maxime pour faire cette manip cotÃ©e api

    if (!acc.withKey[key]) acc.withKey[key] = [];

    acc.withKey[key].push(curr);

    return acc;
};

export const getLastVersementsDate = versements => {
    const orderedVersements = versements.sort((versementA, versementB) => {
        const dateA = new Date(versementA.dateOperation);
        const dateB = new Date(versementB.dateOperation);

        return dateB.getTime() - dateA.getTime();
    });

    if (!orderedVersements.length) return null;

    return new Date(orderedVersements[0].dateOperation);
};

export const sortByPath = (elements, path) => {
    return elements.sort((elementA, elementB) => {
        let attributeA;
        let attributeB;

        if (path.includes("lastDate")) {
            attributeA = getLastVersementsDate(elementA.versements || []);
            attributeB = getLastVersementsDate(elementB.versements || []);
        } else if (path.includes("amount")) {
            attributeA = elementA.versements?.reduce((acc, versement) => acc + versement.amount, 0);
            attributeB = elementB.versements?.reduce((acc, versement) => acc + versement.amount, 0);
        } else {
            attributeA = lodash.get(elementA, path);
            attributeB = lodash.get(elementB, path);
        }

        if (!attributeA && !attributeB) return sortByDateAsc(elementB, elementA);
        else if (!attributeA) return 1;
        else if (!attributeB) return -1;

        if (typeof attributeA == "number" || !isNaN(Number(attributeA))) {
            return attributeB - attributeA;
        }

        // Check if string is date
        const dateA = new Date(attributeA);
        const dateB = new Date(attributeB);

        if (!isNaN(dateA) && !isNaN(dateB)) {
            return dateB.getTime() - dateA.getTime();
        }

        if (attributeB.toLowerCase() > attributeA.toLowerCase()) return 1;
        if (attributeB.toLowerCase() < attributeA.toLowerCase()) return -1;

        return sortByDateAsc(elementB, elementA);
    });
};
