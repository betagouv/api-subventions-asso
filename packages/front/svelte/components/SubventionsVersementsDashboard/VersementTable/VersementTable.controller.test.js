import * as dataHelper from "../../../helpers/dataHelper";
jest.mock("../../../helpers/dataHelper");
import * as dateHelper from "../../../helpers/dateHelper";
jest.mock("../../../helpers/dateHelper");
import * as Helper from "../helper";
jest.mock("../helper");
import VersementTableController from "./VersementTable.controller";

describe("VersementTableController", () => {
    const ELEMENT = {
        versements: [{ centreFinancier: "CENTRE A" }]
    };

    describe("_extractTableDataFromElement()", () => {
        const mockCountTotalVersement = jest.spyOn(VersementTableController, "countTotalVersement");
        beforeAll(() => mockCountTotalVersement.mockImplementation(jest.fn()));
        afterEach(() => mockCountTotalVersement.mockClear());
        afterAll(() => mockCountTotalVersement.mockRestore());

        it("should return an object with properties", () => {
            const actual = Object.keys(VersementTableController._extractTableDataFromElement(ELEMENT));
            expect(actual).toEqual(["totalAmount", "centreFinancier", "lastVersementDate"]);
        });

        it("should return an array of values", () => {
            const actual = VersementTableController._extractTableDataFromElement(ELEMENT, true);
            expect(actual).toEqual([undefined, undefined, undefined]);
        });

        it("should call getLastVersementsDate()", () => {
            VersementTableController._extractTableDataFromElement(ELEMENT);
            expect(Helper.getLastVersementsDate).toHaveBeenCalledTimes(1);
        });

        it("should call countTotalVersement()", () => {
            VersementTableController._extractTableDataFromElement(ELEMENT);
            expect(VersementTableController.countTotalVersement).toHaveBeenCalledTimes(1);
        });

        it("should call valueOrHyphen() twice", () => {
            VersementTableController._extractTableDataFromElement(ELEMENT);
            expect(dataHelper.valueOrHyphen).toHaveBeenCalledTimes(2);
        });

        it("should call withTwoYearDigit() ", () => {
            VersementTableController._extractTableDataFromElement(ELEMENT);
            expect(dateHelper.withTwoDigitYear).toHaveBeenCalledTimes(1);
        });
    });

    describe("extractHeaders()", () => {
        it("return an array of header", () => {
            const actual = VersementTableController.extractHeaders();
            expect(actual).toMatchSnapshot();
        });
    });

    describe("extractRows()", () => {
        const mockExtractTableDataFromElement = jest.spyOn(VersementTableController, "_extractTableDataFromElement");
        beforeAll(() => mockExtractTableDataFromElement.mockImplementation(jest.fn()));
        afterAll(() => mockExtractTableDataFromElement.mockRestore());
        it("should call _extractTableDataFromElement for each element in array", () => {
            VersementTableController.extractRows([{ versements: [] }, { versements: [] }]);
            expect(mockExtractTableDataFromElement).toHaveBeenCalledTimes(2);
        });
        it("should not call _extractTableDataFromElement if element has no versements", () => {
            VersementTableController.extractRows([{}, { versements: [] }]);
            expect(mockExtractTableDataFromElement).toHaveBeenCalledTimes(1);
        });

        it("should return an array", () => {
            const expected = [null, null];
            const actual = VersementTableController.extractRows([{}, {}]);
            expect(actual).toEqual(expected);
        });
    });
});
