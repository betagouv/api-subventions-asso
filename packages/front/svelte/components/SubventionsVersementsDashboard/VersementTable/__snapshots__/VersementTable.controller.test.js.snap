// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VersementTableController extractHeaders() return an array of header 1`] = `
[
  "Montant vers√©",
  "Centre financier",
  "Date du versement",
]
`;
